File: lib/java/src/main/java/io/github/telepact/internal/binary/ServerBase64Encode.java
//|
//|  Copyright The Telepact Authors
//|
//|  Licensed under the Apache License, Version 2.0 (the "License");
//|  you may not use this file except in compliance with the License.
//|  You may obtain a copy of the License at
//|
//|  https://www.apache.org/licenses/LICENSE-2.0
//|
//|  Unless required by applicable law or agreed to in writing, software
//|  distributed under the License is distributed on an "AS IS" BASIS,
//|  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//|  See the License for the specific language governing permissions and
//|  limitations under the License.
//|

package io.github.telepact.internal.binary;

import java.util.Base64;
import java.util.List;
import java.util.Map;

public class ServerBase64Encode {

    public static void serverBase64Encode(List<Object> message) {
        Map<String, Object> headers = (Map<String, Object>) message.get(0);
        Map<String, Object> body = (Map<String, Object>) message.get(1);

        Map<String, Object> base64Paths = (Map<String, Object>) headers.getOrDefault("@base64_", Map.of());

        travelBase64Encode(body, base64Paths);
    }

    private static Object travelBase64Encode(Object value, Object base64Paths) {
        if (base64Paths instanceof Map) {
            Map<String, Object> base64PathsMap = (Map<String, Object>) base64Paths;
            for (Map.Entry<String, Object> entry : base64PathsMap.entrySet()) {
                String key = entry.getKey();
                Object val = entry.getValue();

                if (Boolean.TRUE.equals(val)) {
                    if ("*".equals(key) && value instanceof List) {
                        List<Object> valueList = (List<Object>) value;
                        for (int i = 0; i < valueList.size(); i++) {
                            valueList.set(i, travelBase64Encode(valueList.get(i), val));
                        }
                    } else if ("*".equals(key) && value instanceof Map) {
                        Map<String, Object> valueMap = (Map<String, Object>) value;
                        for (Map.Entry<String, Object> valueEntry : valueMap.entrySet()) {
                            valueMap.put(valueEntry.getKey(), travelBase64Encode(valueEntry.getValue(), val));
                        }
                    } else if (value instanceof Map) {
                        Map<String, Object> valueMap = (Map<String, Object>) value;
                        valueMap.put(key, travelBase64Encode(valueMap.get(key), val));
                    } else {
                        throw new IllegalArgumentException("Invalid base64 path: " + key + " for value: " + value);
                    }
                } else {
                    if ("*".equals(key) && value instanceof List) {
                        List<Object> valueList = (List<Object>) value;
                        for (Object v : valueList) {
                            travelBase64Encode(v, val);
                        }
                    } else if ("*".equals(key) && value instanceof Map) {
                        Map<String, Object> valueMap = (Map<String, Object>) value;
                        for (Object v : valueMap.values()) {
                            travelBase64Encode(v, val);
                        }
                    } else if (value instanceof Map) {
                        Map<String, Object> valueMap = (Map<String, Object>) value;
                        travelBase64Encode(valueMap.get(key), val);
                    } else {
                        throw new IllegalArgumentException("Invalid base64 path: " + key + " for value: " + value);
                    }
                }
            }
            return null;
        } else if (Boolean.TRUE.equals(base64Paths) && (value instanceof byte[] || value == null)) {
            if (value == null) {
                return null;
            }
            return Base64.getEncoder().encodeToString((byte[]) value);
        } else {
            throw new IllegalArgumentException("Invalid base64 path: " + base64Paths + " for value: " + value);
        }
    }
}

File: lib/go/internal/binary/ServerBase64Encode.go
//|
//|  Copyright The Telepact Authors
//|
//|  Licensed under the Apache License, Version 2.0 (the "License");
//|  you may not use this file except in compliance with the License.
//|  You may obtain a copy of the License at
//|
//|  https://www.apache.org/licenses/LICENSE-2.0
//|
//|  Unless required by applicable law or agreed to in writing, software
//|  distributed under the License is distributed on an "AS IS" BASIS,
//|  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//|  See the License for the specific language governing permissions and
//|  limitations under the License.
//|

package binary

import (
	"encoding/base64"
	"fmt"
)

// ServerBase64Encode traverses the message body and encodes byte arrays into base64 strings based on the provided paths.
func ServerBase64Encode(message []any) error {
	if len(message) < 2 {
		return fmt.Errorf("invalid message: expected headers and body, got %d elements", len(message))
	}

	headers, err := ensureStringMap(message[0])
	if err != nil {
		return err
	}

	body, err := ensureStringKeyedMap(message[1])
	if err != nil {
		return err
	}

	base64Paths := headers["@base64_"]
	if base64Paths == nil {
		return nil
	}

	_, err = travelServerBase64Encode(body, base64Paths)
	return err
}

func travelServerBase64Encode(value any, base64Paths any) (any, error) {
	switch paths := base64Paths.(type) {
	case map[string]any:
		for key, descriptor := range paths {
			if err := applyBase64EncodeDescriptor(value, key, descriptor); err != nil {
				return nil, err
			}
		}
		return value, nil
	case map[any]any:
		for rawKey, descriptor := range paths {
			key := fmt.Sprint(rawKey)
			if err := applyBase64EncodeDescriptor(value, key, descriptor); err != nil {
				return nil, err
			}
		}
		return value, nil
	case bool:
		if !paths {
			return nil, fmt.Errorf("invalid base64 path: %v for value: %v", base64Paths, value)
		}
		if value == nil {
			return nil, nil
		}
		bytesValue, ok := value.([]byte)
		if !ok {
			return nil, fmt.Errorf("invalid base64 path: %v for value: %T", base64Paths, value)
		}
		return base64.StdEncoding.EncodeToString(bytesValue), nil
	default:
		return nil, fmt.Errorf("invalid base64 path: %v for value: %v", base64Paths, value)
	}
}

func applyBase64EncodeDescriptor(container any, key string, descriptor any) error {
	if boolDescriptor, ok := descriptor.(bool); ok {
		if !boolDescriptor {
			return fmt.Errorf("invalid base64 path: %v for value: %v", key, container)
		}
		return applyBase64EncodeLeaf(container, key)
	}

	return applyBase64EncodeNested(container, key, descriptor)
}

func applyBase64EncodeLeaf(container any, key string) error {
	switch typed := container.(type) {
	case []any:
		if key != "*" {
			return fmt.Errorf("invalid base64 path: %s for value: %v", key, container)
		}
		for i, item := range typed {
			encoded, err := travelServerBase64Encode(item, true)
			if err != nil {
				return err
			}
			typed[i] = encoded
		}
		return nil
	case map[string]any:
		if key == "*" {
			for childKey, item := range typed {
				encoded, err := travelServerBase64Encode(item, true)
				if err != nil {
					return err
				}
				typed[childKey] = encoded
			}
			return nil
		}
		child, ok := typed[key]
		if !ok {
			return fmt.Errorf("invalid base64 path: %s for value: %v", key, container)
		}
		encoded, err := travelServerBase64Encode(child, true)
		if err != nil {
			return err
		}
		typed[key] = encoded
		return nil
	case map[any]any:
		if key == "*" {
			for childKey, item := range typed {
				encoded, err := travelServerBase64Encode(item, true)
				if err != nil {
					return err
				}
				typed[childKey] = encoded
			}
			return nil
		}
		child, ok := typed[key]
		if !ok {
			return fmt.Errorf("invalid base64 path: %s for value: %v", key, container)
		}
		encoded, err := travelServerBase64Encode(child, true)
		if err != nil {
			return err
		}
		typed[key] = encoded
		return nil
	default:
		return fmt.Errorf("invalid base64 path: %s for value: %v", key, container)
	}
}

func applyBase64EncodeNested(container any, key string, descriptor any) error {
	switch typed := container.(type) {
	case []any:
		if key != "*" {
			return fmt.Errorf("invalid base64 path: %s for value: %v", key, container)
		}
		for _, item := range typed {
			if _, err := travelServerBase64Encode(item, descriptor); err != nil {
				return err
			}
		}
		return nil
	case map[string]any:
		if key == "*" {
			for childKey, item := range typed {
				if updated, err := travelServerBase64Encode(item, descriptor); err != nil {
					return err
				} else if updated != nil {
					typed[childKey] = updated
				}
			}
			return nil
		}
		child, ok := typed[key]
		if !ok {
			return fmt.Errorf("invalid base64 path: %s for value: %v", key, container)
		}
		if updated, err := travelServerBase64Encode(child, descriptor); err != nil {
			return err
		} else if updated != nil {
			typed[key] = updated
		}
		return nil
	case map[any]any:
		if key == "*" {
			for childKey, item := range typed {
				if updated, err := travelServerBase64Encode(item, descriptor); err != nil {
					return err
				} else if updated != nil {
					typed[childKey] = updated
				}
			}
			return nil
		}
		child, ok := typed[key]
		if !ok {
			return fmt.Errorf("invalid base64 path: %s for value: %v", key, container)
		}
		if updated, err := travelServerBase64Encode(child, descriptor); err != nil {
			return err
		} else if updated != nil {
			typed[key] = updated
		}
		return nil
	default:
		return fmt.Errorf("invalid base64 path: %s for value: %v", key, container)
	}
}

func ensureStringKeyedMap(value any) (map[string]any, error) {
	switch typed := value.(type) {
	case map[string]any:
		return typed, nil
	case map[any]any:
		result := make(map[string]any, len(typed))
		for key, val := range typed {
			result[fmt.Sprint(key)] = val
		}
		return result, nil
	default:
		return nil, fmt.Errorf("expected map[string]any, got %T", value)
	}
}
